protocol/consistencychecks.go

```diff
@@ -0,0 +1,298 @@
+			proofType == proofOfAbsence {
+			return PassedWithAProofOfAbsence
+		}
+		return cc.verifyProofTypeUsingPromises(msg.Error, proofType)
+
+	case KeyLookupType:
+		if msg.Error == ErrorNameNotFound &&
+			proofType == proofOfAbsence {
+			return PassedWithAProofOfAbsence
+		} else if msg.Error == Success &&
+			proofType == proofOfInclusion {
+			return PassedWithAProofOfInclusion
+		}
+		return cc.verifyProofTypeUsingPromises(msg.Error, proofType)
+
+	case MonitoringType:
+		if proofType == proofOfAbsence {
+			return ErrorBadProofType
+		}
+		return PassedWithAProofOfInclusion
+
+	case KeyLookupInEpochType:
+		// TODO:
+
+	}
+
+	// this return should be unreachable
+	// unless the developers mistakenly passed an invalid request type.
+	return ErrorMalformedClientMessage
+}
+
+func (cc *ConsistencyChecks) verifyDirectoryProof(df *DirectoryProof,
+	uname string, key []byte) (proofType int, verifResult error) {
+
+	verifResult = nil
+	proofType = invalidProof
+
+	ap := df.AP
+	str := df.STR
+	tb := df.TB
+
+	if verifResult = cc.verifySTR(str); verifResult != nil {
+		return
+	}
+
+	if proofType, verifResult = cc.verifyAuthPath(uname, key,
+		ap, str); verifResult != nil {
+		return
+	}
+
+	if verifResult = cc.verifyReturnedPromise(tb, uname, key, str, ap); verifResult != nil {
+		return
+	}
+
+	if verifResult = cc.verifyFulfilledPromise(uname, str, ap); verifResult != nil {
+		return
+	}
+
+	// re-assign new state for the client
+	if str.Epoch == cc.CurrentEpoch+1 {
+		cc.CurrentEpoch++
+		cc.SavedSTR = str.Signature
+	}
+	return
+}
+
+func (cc *ConsistencyChecks) verifyDirectoryProofs(dfs *DirectoryProofs,
+	uname string, key []byte) (int, error) {
+	// TODO: implement verifications for multi ranges of epoch
+	return invalidProof, nil
+}
+
+func (cc *ConsistencyChecks) verifyAuthPath(uname string, key []byte,
+	ap *m.AuthenticationPath,
+	str *m.SignedTreeRoot) (int, error) {
+	proofType := proofOfAbsence
+
+	// verify VRF Index
+	vrfKey := vrf.PublicKey(str.Policies.VrfPublicKey)
+	if !vrfKey.Verify([]byte(uname), ap.LookupIndex, ap.VrfProof) {
+		return invalidProof, ErrorBadVRFProof
+	}
+
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) { // proof of inclusion */
+		// verify name-to-key binding
+		// key is nil when the user does lookup for the first time
+		if key != nil && !ap.VerifyBinding(key) {
+			return invalidProof, ErrorBadBinding
+		}
+		// verify commitment
+		if !ap.Leaf.Commitment.Verify([]byte(uname), ap.Leaf.Value) {
+			return invalidProof, ErrorBadCommitment
+		}
+		proofType = proofOfInclusion
+	}
+
+	// verify auth path
+	if !ap.Verify(str.TreeHash) {
+		return invalidProof, ErrorBadAuthPath
+	}
+
+	return proofType, nil
+}
+
+func (cc *ConsistencyChecks) verifySTR(str *m.SignedTreeRoot) error {
+	// verify STR's signature
+	if !cc.signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify hash chain
+	if (str.Epoch == cc.CurrentEpoch && bytes.Equal(cc.SavedSTR, str.Signature)) ||
```

Comment:
I feel like these conditions shouldn't be buried down here.  It makes it hard to ascertain whether we've actually verified our expectations.